clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -985,6 +1018,60 @@ private Record nextFetchedRecord() {
+            *   0. Check whether the pid is in the 'abortedProducerIds' set && the entry does not include an abort marker.
+            *      If so, skip the entry.
+            *   1. If the pid is in aborted pids and the entry contains an abort marker, remove the pid from
+            *      aborted pids and skip the entry.
+            *   2. Check lowest offset entry in the abort index. If the PID of the current entry matches the
+            *      pid of the abort index entry, and the incoming offset is no smaller than the abort index offset,
+            *      this means that the entry has been aborted. Add the pid to the aborted pids set, and remove
+            *      the entry from the abort index.
+            */
+            FetchResponse.AbortedTransaction nextAbortedTransaction = abortedTransactions.peek();
+            if (abortedProducerIds.contains(batch.producerId())
+                    || (nextAbortedTransaction != null && nextAbortedTransaction.producerId == batch.producerId() && nextAbortedTransaction.firstOffset <= batch.baseOffset())) {
+                if (abortedProducerIds.contains(batch.producerId()) && containsAbortMarker(batch)) {
+                    abortedProducerIds.remove(batch.producerId());
+                } else if (nextAbortedTransaction != null && nextAbortedTransaction.producerId == batch.producerId() && nextAbortedTransaction.firstOffset <= batch.baseOffset()) {
+                    abortedProducerIds.add(batch.producerId());
+                    abortedTransactions.poll();
+                }
+                log.trace("Skipping aborted record batch with producerId {} and base offset {}", batch.producerId(), batch.baseOffset());
+                return true;
+            }
+            return false;
+        }
+
+        private PriorityQueue<FetchResponse.AbortedTransaction> abortedTransactions(FetchResponse.PartitionData partition) {
+            PriorityQueue<FetchResponse.AbortedTransaction> abortedTransactions = null;
+            if (partition.abortedTransactions != null && !partition.abortedTransactions.isEmpty()) {
+                abortedTransactions = new PriorityQueue<>(
+                        partition.abortedTransactions.size(),
+                        new Comparator<FetchResponse.AbortedTransaction>() {
+                            @Override
+                            public int compare(FetchResponse.AbortedTransaction o1, FetchResponse.AbortedTransaction o2) {
+                                return Long.compare(o1.firstOffset, o2.firstOffset);
+                            }
+                        }
+                );
+                abortedTransactions.addAll(partition.abortedTransactions);
+            } else {
+                abortedTransactions = new PriorityQueue<>();
+            }
+            return abortedTransactions;
+        }
+
+        private boolean containsAbortMarker(RecordBatch batch) {
```

Comment:
When the abort index patch comes in, we can leverage the `CONTROL_SEQUENCE` value to only look inside the batch there is a control message. This way we won't have to open all batches, potentially decompressing messages.